File: mindspore_serving/worker/register/postprocess.py
# Copyright 2020 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Postprocessing registration interface"""

from mindspore_serving._mindspore_serving import PostprocessStorage_
from mindspore_serving.worker.common import get_servable_dir, get_func_name
from mindspore_serving import log as logger


def check_postprocess(postprocess_name, inputs_count, outputs_count):
    """Check whether inputs and outputs count is equal with last registered"""
    postprocess_info = PostprocessStorage_.get_instance().get_pycpp_postprocess_info(postprocess_name)
    if not postprocess_info:
        return
    last_inputs_count, last_output_count = postprocess_info
    if last_inputs_count != inputs_count:
        raise RuntimeError(f" Postprocess '{postprocess_name}' inputs count {inputs_count} not match "
                           f"last registered count {last_inputs_count}")
    if last_output_count != outputs_count:
        raise RuntimeError(f" Postprocess '{postprocess_name}' outputs count {outputs_count} not match "
                           f"last registered count {last_output_count}")


class PostprocessStorage:
    """Register and get postprocess info, postprocess info include: func, name, input and output count"""

    def __init__(self):
        self.postprocess = {}
        self.storage = PostprocessStorage_.get_instance()

    def clear(self):
        self.postprocess = {}

    def register(self, fun, postprocess_name, inputs_count, outputs_count):
        check_postprocess(postprocess_name, inputs_count, outputs_count)
        self.postprocess[postprocess_name] = {"fun": fun, "inputs_count": inputs_count, "outputs_count": outputs_count}
        self.storage.register(postprocess_name, inputs_count, outputs_count)

    def get(self, postprocess_name):
        postprocess = self.postprocess.get(postprocess_name, None)
        if postprocess is None:
            raise RuntimeError(f"Postprocess '{postprocess_name}' not found")
        return postprocess


postprocess_storage = PostprocessStorage()


def register_postprocess(func, inputs_count, outputs_count):
    """register postprocess"""
    servable_name = get_servable_dir()
    func_name = get_func_name(func)
    name = servable_name + "." + func_name

    logger.info(f"Register postprocess {name} {inputs_count} {outputs_count}")
    postprocess_storage.register(func, name, inputs_count, outputs_count)

File: mindspore_serving/ccsrc/worker/postprocess.cc
/**
 * Copyright 2020 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "worker/postprocess.h"
#include <utility>

namespace mindspore::serving {

bool PostprocessStorage::Register(const std::string &postprocess_name, std::shared_ptr<PostprocessBase> postprocess) {
  if (postprocess_map_.find(postprocess_name) != postprocess_map_.end()) {
    MSI_LOG_WARNING << "postprocess " << postprocess_name << " has been registered";
    return false;
  }
  postprocess_map_[postprocess_name] = std::move(postprocess);
  return true;
}

void PostprocessStorage::Unregister(const std::string &postprocess_name) {
  auto it = postprocess_map_.find(postprocess_name);
  if (it == postprocess_map_.end()) {
    return;
  }
  postprocess_map_.erase(it);
}

std::shared_ptr<PostprocessBase> PostprocessStorage::GetPostprocess(const std::string &postprocess_name) const {
  auto it = postprocess_map_.find(postprocess_name);
  if (it != postprocess_map_.end()) {
    return it->second;
  }
  return nullptr;
}

PostprocessStorage &PostprocessStorage::Instance() {
  static PostprocessStorage storage;
  return storage;
}

RegPostprocess::RegPostprocess(const std::string &postprocess_name, std::shared_ptr<PostprocessBase> postprocess) {
  postprocess_name_ = postprocess_name;
  MSI_LOG_INFO << "Register C++ postprocess " << postprocess_name;
  register_success_ = PostprocessStorage::Instance().Register(postprocess_name, std::move(postprocess));
}

RegPostprocess::~RegPostprocess() {
  if (register_success_) {
    MSI_LOG_INFO << "Unregister C++ preprocess " << postprocess_name_;
    PostprocessStorage::Instance().Unregister(postprocess_name_);
  }
}

}  // namespace mindspore::serving
